File: studio/windows/AuraStudio.Core/Networking/AppUpdateDecision.cs
namespace AuraStudio.Core.Networking;

/// <summary>
/// Para qué plataforma se busca el instalador (ST-193, port literal de
/// <c>AppUpdatePlatform</c> de Swift).
///
/// <para>Las tres están acá aunque Windows solo use dos, y es deliberado: el
/// patrón de nombres de los assets es <b>el único contrato nuevo</b> de esta
/// funcionalidad, lo congeló la sesión maestra para las dos plataformas, y
/// tenerlo escrito en un solo lugar por lado es lo que hace que los dos digan lo
/// mismo.</para>
///
/// <para>El patrón, tal como quedó fijado (y tal como ya lo cumplen los Releases
/// v0.2.2 y v0.2.3):</para>
/// <code>
/// tag:  v&lt;versión&gt;
/// mac:  AuraStudio-&lt;versión&gt;.dmg
/// win:  AuraStudioSetup-&lt;versión&gt;-arm64.exe
///       AuraStudioSetup-&lt;versión&gt;-x64.exe
/// </code>
///
/// <para><b>Ningún otro asset cuenta.</b> En Windows, elegir mal la arquitectura
/// es peor que no ofrecer nada: ST-135 documenta que el Setup x64 en una máquina
/// ARM avisa y deja continuar, así que ofrecerlo por defecto sería empujar al
/// usuario a la versión lenta.</para>
/// </summary>
public enum AppUpdatePlatform
{
    Mac,
    WindowsArm64,
    WindowsX64
}

public static class AppUpdatePlatformNames
{
    /// <summary>
    /// La de este proceso. Se mira la arquitectura del <b>proceso</b> y no la de
    /// la máquina: un Aura Studio x64 corriendo bajo emulación en un Windows ARM
    /// tiene que ofrecerse su propio instalador.
    /// </summary>
    public static AppUpdatePlatform Current =>
        System.Runtime.InteropServices.RuntimeInformation.ProcessArchitecture
            == System.Runtime.InteropServices.Architecture.Arm64
            ? AppUpdatePlatform.WindowsArm64
            : AppUpdatePlatform.WindowsX64;

    /// <summary>El nombre EXACTO del asset. La versión va sin la <c>v</c> del tag.</summary>
    public static string AssetName(this AppUpdatePlatform platform, SemVer version) => platform switch
    {
        AppUpdatePlatform.Mac => $"AuraStudio-{version.ReleaseString}.dmg",
        AppUpdatePlatform.WindowsArm64 => $"AuraStudioSetup-{version.ReleaseString}-arm64.exe",
        _ => $"AuraStudioSetup-{version.ReleaseString}-x64.exe"
    };
}

/// <summary>Lo que hay que mostrar cuando hay una versión más nueva de la app.</summary>
/// <param name="Version">La versión publicada, ya parseada.</param>
/// <param name="Tag">
/// El tag tal cual lo publicó GitHub (<c>v0.3.0</c>) — es lo que se persiste para
/// no repetir el aviso.
/// </param>
/// <param name="ReleasePageUrl">La página del Release, para "Ver novedades".</param>
/// <param name="Asset">
/// El asset a bajar para ESTA plataforma, o <c>null</c> si el Release no lo trae.
/// </param>
/// <param name="AssetName">El nombre del asset esperado, para poder decirlo si falta.</param>
public readonly record struct AppUpdateAvailable(
    SemVer Version,
    string Tag,
    string ReleasePageUrl,
    GitHubReleaseAsset? Asset,
    string AssetName)
{
    /// <summary>La URL pública de descarga, o <c>null</c> si el asset no está.</summary>
    public string? DownloadUrl => Asset?.BrowserDownloadUrl is { Length: > 0 } url ? url : null;

    /// <summary>
    /// Si se puede ofrecer el botón de descargar. Sin asset se anuncia igual la
    /// versión, pero con el enlace a la página: un botón "Descargar" que falla es
    /// peor que no tenerlo.
    /// </summary>
    public bool CanDownload => Asset is not null;
}

/// <summary>
/// ST-193 (port literal desde Swift, que es la referencia por acuerdo de la
/// sesión maestra): decidir si hay una versión más nueva <b>de Aura Studio</b> y
/// cuál archivo le corresponde a esta plataforma.
///
/// <para>Es <b>pura a propósito</b>: no consulta la red, no lee el ensamblado, no
/// toca disco. Se le dan la versión instalada y los Releases ya traídos, y
/// devuelve una decisión. Por eso se prueba entera sin red, y por eso las dos
/// plataformas pueden compartir exactamente las mismas reglas.</para>
///
/// <para><b>Las reglas, en orden:</b></para>
/// <list type="number">
/// <item>Un <b>draft</b> nunca cuenta: no es instalable.</item>
/// <item>Una <b>prerelease</b> cuenta solo si se piden. Hoy todo lo publicado es
/// beta, así que en la práctica cuenta siempre; el interruptor importa el día
/// que exista un canal estable.</item>
/// <item>El tag se lee como SemVer (<c>v0.3.0</c> → <c>0.3.0</c>). Un tag que no
/// parsea <b>se ignora</b>, no rompe nada.</item>
/// <item>De los que quedan, gana el <b>mayor</b>, no el primero.</item>
/// <item>Solo hay novedad si ese mayor es <b>estrictamente mayor</b> que lo
/// instalado. Nunca se ofrece "actualizar" hacia atrás.</item>
/// <item>El asset se busca por <b>nombre exacto</b>. Si no está, igual se avisa
/// de la versión nueva, pero sin botón de descarga.</item>
/// </list>
/// </summary>
public static class AppUpdateDecision
{
    /// <summary>
    /// El repositorio donde se publican los instaladores de la app — distinto
    /// del del firmware (<c>Aura-Firmware</c> y hermanos).
    /// </summary>
    public const string Repository = "Ricolinos/Aura-Studio";

    /// <summary>
    /// Decide. <paramref name="installedVersion"/> es la versión de la app
    /// corriendo; <paramref name="releases"/> es lo que devolvió GitHub.
    ///
    /// <para>Devuelve <c>null</c> cuando no hay nada que ofrecer — incluido el
    /// caso de una versión instalada que no parsea, donde lo prudente es callar:
    /// sin saber qué hay instalado no se puede afirmar que algo sea más
    /// nuevo.</para>
    /// </summary>
    public static AppUpdateAvailable? Decide(
        string installedVersion,
        IReadOnlyList<GitHubRelease> releases,
        bool includePrereleases,
        AppUpdatePlatform platform)
    {
        if (SemVer.Parse(installedVersion) is not { } installed) return null;

        if (GitHubReleaseChecker.PickLatest(releases, includePrereleases) is not { } latest) return null;
        if (SemVer.Parse(latest.TagName) is not { } latestVersion) return null;
        if (installed >= latestVersion) return null;

        string assetName = platform.AssetName(latestVersion);

        return new AppUpdateAvailable(
            latestVersion,
            latest.TagName,
            ReleasePageUrl(latest),
            latest.Asset(assetName),
            assetName);
    }

    /// <summary>
    /// La página del Release. Se prefiere la que da GitHub; si no vino
    /// (respuesta recortada, caché viejo), se arma con el repo y el tag — es una
    /// URL estable y documentada de GitHub.
    /// </summary>
    public static string ReleasePageUrl(GitHubRelease release) =>
        release.HtmlUrl is { Length: > 0 } html
            ? html
            : $"https://github.com/{Repository}/releases/tag/{release.TagName}";
}

File: studio/AuraStudio/Sources/AuraStudio/Models/AppUpdateDecision.swift
import Foundation

/// ST-193: decidir si hay una versión más nueva **de Aura Studio** y
/// cuál archivo le corresponde a esta plataforma.
///
/// Es la pieza nueva de la propuesta de ST-191, y es **pura a
/// propósito**: no consulta la red, no lee el bundle, no toca disco. Se
/// le dan la versión instalada y los Releases ya traídos, y devuelve una
/// decisión. Por eso se puede probar entera sin red y por eso sirve como
/// **referencia para Windows**, que la porta citando esta ST (acuerdo de
/// la sesión maestra: Swift es la referencia).
///
/// ## Las reglas, en orden
///
/// 1. Un **draft** nunca cuenta: no es instalable.
/// 2. Una **prerelease** cuenta solo si `includePrereleases`. Hoy todo
///    lo publicado es beta, así que en la práctica cuenta siempre; el
///    interruptor importa el día que exista un canal estable.
/// 3. El tag se lee como SemVer (`v0.3.0` → `0.3.0`). Un tag que no
///    parsea **se ignora**, no rompe nada.
/// 4. De los que quedan, gana el mayor.
/// 5. Solo hay novedad si ese mayor es **estrictamente mayor** que lo
///    instalado. Nunca se ofrece "actualizar" hacia atrás.
/// 6. El asset se busca por **nombre exacto**, según el patrón congelado
///    (ver `AppUpdatePlatform.assetName`). Si no está, **igual se avisa
///    de la versión nueva**, pero sin botón de descarga: se enlaza la
///    página del Release. Un botón "Descargar" que falla es peor que no
///    tenerlo.
enum AppUpdateDecision {
    /// El repositorio donde se publican los instaladores de la app --
    /// distinto del del firmware (`Aura-Firmware` y hermanos).
    static let repository = "Ricolinos/Aura-Studio"

    /// Lo que hay que mostrar cuando hay una versión más nueva.
    struct Available: Equatable {
        /// La versión publicada, ya parseada.
        let version: SemVer
        /// El tag tal cual lo publicó GitHub (`v0.3.0`) -- es lo que se
        /// persiste para no repetir el aviso.
        let tag: String
        /// La página del Release, para "Ver novedades".
        let releasePageURL: URL?
        /// El archivo a bajar para ESTA plataforma. `nil` si el Release
        /// no trae el asset esperado.
        let downloadURL: URL?
        /// El nombre del asset esperado, para poder decirlo si falta.
        let assetName: String
    }

    /// El resultado de una comprobación, para la que el usuario pidió a
    /// mano. El chequeo automático solo mira `.available`.
    enum Outcome: Equatable {
        case available(Available)
        /// Consultado con éxito: no hay nada más nuevo.
        case upToDate
        /// No se pudo preguntar. **No es lo mismo que "no hay
        /// novedades"** y la UI no puede decirlo igual -- es el defecto
        /// que Windows arregló en ST-210 para el chequeo del firmware.
        case couldNotCheck(String)
    }

    /// Decide. `installedVersion` es la versión de la app corriendo
    /// (`AppVersion.current`); `releases` es lo que devolvió GitHub.
    ///
    /// Devuelve `nil` cuando no hay nada que ofrecer -- incluido el caso
    /// de una versión instalada que no parsea, donde lo prudente es
    /// callar: sin saber qué hay instalado no se puede afirmar que algo
    /// sea más nuevo.
    static func decide(installedVersion: String,
                       releases: [GitHubRelease],
                       includePrereleases: Bool,
                       platform: AppUpdatePlatform = .current) -> Available? {
        guard let installed = SemVer.parse(installedVersion) else { return nil }
        guard let latest = GitHubReleaseChecker.pickLatest(from: releases,
                                                          includePrereleases: includePrereleases),
              let latestVersion = SemVer.parse(latest.tagName),
              installed < latestVersion else {
            return nil
        }

        let assetName = platform.assetName(forVersion: latestVersion)
        let download = latest.asset(named: assetName)?.browserDownloadURL.flatMap(URL.init(string:))

        return Available(version: latestVersion,
                         tag: latest.tagName,
                         releasePageURL: releasePageURL(for: latest),
                         downloadURL: download,
                         assetName: assetName)
    }

    /// La página del Release. Se prefiere la que da GitHub; si no vino
    /// (respuesta recortada, caché viejo), se arma con el repo y el tag
    /// -- es una URL estable y documentada de GitHub.
    static func releasePageURL(for release: GitHubRelease) -> URL? {
        if let html = release.htmlURL, let url = URL(string: html) { return url }
        return URL(string: "https://github.com/\(repository)/releases/tag/\(release.tagName)")
    }
}


/// ST-193: para qué plataforma se busca el instalador.
///
/// Las tres están acá aunque macOS solo use una, y es deliberado: el
/// patrón de nombres de los assets es **el único contrato nuevo** de
/// esta funcionalidad, lo congeló la sesión maestra para las dos
/// plataformas, y tenerlo escrito en un solo lugar es lo que hace que
/// Windows pueda portarlo sin volver a decidir nada.
///
/// El patrón, tal como quedó fijado (y tal como ya lo cumplen los
/// Releases v0.2.2 y v0.2.3):
///
/// ```
/// tag:  v<versión>
/// mac:  AuraStudio-<versión>.dmg
/// win:  AuraStudioSetup-<versión>-arm64.exe
///       AuraStudioSetup-<versión>-x64.exe
/// ```
///
/// **Ningún otro asset cuenta.** En Windows, elegir mal la arquitectura
/// es peor que no ofrecer nada: ST-135 documenta que el Setup x64 en una
/// máquina ARM avisa y deja continuar, así que ofrecerlo por defecto
/// sería empujar al usuario a la versión lenta.
enum AppUpdatePlatform: Equatable, Sendable {
    case mac
    case windowsARM64
    case windowsX64

    /// La de esta app. En macOS siempre `.mac` -- el `.dmg` es
    /// universal, no hay variante por arquitectura.
    static var current: AppUpdatePlatform { .mac }

    /// El nombre EXACTO del asset. La versión va sin la `v` del tag.
    func assetName(forVersion version: SemVer) -> String {
        switch self {
        case .mac: return "AuraStudio-\(version.releaseString).dmg"
        case .windowsARM64: return "AuraStudioSetup-\(version.releaseString)-arm64.exe"
        case .windowsX64: return "AuraStudioSetup-\(version.releaseString)-x64.exe"
        }
    }
}

extension SemVer {
    /// La versión como aparece en el NOMBRE de un asset y en el tag
    /// (sin la `v`): `0.3.0`, o `0.3.0-beta` si tuviera sufijo.
    var releaseString: String {
        let core = "\(major).\(minor).\(patch)"
        return prerelease.map { "\(core)-\($0)" } ?? core
    }
}

/// ST-193: la versión de ESTA app.
///
/// La app nunca había necesitado saber su propia versión, así que no
/// había de dónde leerla. Sale de `CFBundleShortVersionString`, que ya
/// se sube en cada release junto con `MARKETING_VERSION` y las dos de
/// Windows (regla de `CLAUDE.md` § Releases: los tres lugares a la vez,
/// en el mismo commit).
///
/// `overrideForTesting` existe porque bajo `swift test` no hay bundle de
/// app: el Info.plist con esa clave solo está en el `.app` que arma
/// `xcodebuild`. Sin esto, cualquier prueba que toque esto mediría el
/// bundle del ejecutable de pruebas, que no tiene versión.
enum AppVersion {
    nonisolated(unsafe) static var overrideForTesting: String?

    static var current: String {
        if let overrideForTesting { return overrideForTesting }
        return Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "0.0.0"
    }
}
